File: Stripe/STPPinManagementService.m
//
//  STPAPIClient+PinManagement.m
//  Stripe
//
//  Created by Arnaud Cavailhez on 4/29/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

#import "STPPinManagementService.h"
#import "STPAPIRequest.h"
#import "STPIssuingCardPin.h"
#import "STPEphemeralKeyManager.h"
#import "STPAPIClient+Private.h"

@interface STPPinManagementService()
@property (nonatomic, strong) STPEphemeralKeyManager *keyManager;
@end

@implementation STPPinManagementService

- (instancetype)initWithKeyProvider:(id<STPIssuingCardEphemeralKeyProvider>)keyProvider {
    self = [super init];
    if (self) {
        _apiClient = [STPAPIClient sharedClient];
        _keyManager = [[STPEphemeralKeyManager alloc] initWithKeyProvider:keyProvider apiVersion:[STPAPIClient apiVersion] performsEagerFetching:NO];
    }
    return self;
}

- (void)retrievePin:(NSString *) cardId
         verificationId:(NSString *) verificationId
        oneTimeCode:(NSString *) oneTimeCode
         completion:(STPPinCompletionBlock) completion{
    NSString *endpoint = [NSString stringWithFormat:@"issuing/cards/%@/pin", cardId];
    NSDictionary *parameters = @{
                                 @"verification": @{
                                         @"id": verificationId,
                                         @"one_time_code": oneTimeCode,
                                         },
                                 };
    [self.keyManager getOrCreateKey:^(STPEphemeralKey * _Nullable ephemeralKey, NSError * _Nullable keyError) {
        if (ephemeralKey == nil) {
            completion(nil, STPPinEphemeralKeyError, keyError);
            return;
        }

        [STPAPIRequest<STPIssuingCardPin *> getWithAPIClient:self.apiClient
                                                    endpoint:endpoint
                                           additionalHeaders:[self.apiClient authorizationHeaderUsingEphemeralKey:ephemeralKey]
                                                  parameters:parameters
                                                deserializer:[STPIssuingCardPin new]
                                                  completion:^(
                                                               STPIssuingCardPin *details,
                                                               __unused     NSHTTPURLResponse *response,
                                                               NSError *error) {
                                                      // Find if there were errors
                                                      if (details.error != nil) {
                                                          NSString* code = details.error[@"code"];
                                                          if ([@"api_key_expired" isEqualToString:code]) {
                                                              completion(nil, STPPinEphemeralKeyError, error);
                                                          } else if ([@"expired" isEqualToString:code]) {
                                                              completion(nil, STPPinErrorVerificationExpired, nil);
                                                          } else if ([@"incorrect_code" isEqualToString:code]) {
                                                              completion(nil, STPPinErrorVerificationCodeIncorrect, nil);
                                                          } else if ([@"too_many_attempts" isEqualToString:code]) {
                                                              completion(nil, STPPinErrorVerificationTooManyAttempts, nil);
                                                          } else if ([@"already_redeemed" isEqualToString:code]) {
                                                              completion(nil, STPPinErrorVerificationAlreadyRedeemed, nil);
                                                          } else {
                                                              completion(nil, STPPinUnknownError, error);
                                                          }
                                                          return;
                                                      }
                                                      completion(details, STPPinSuccess, nil);
                                                  }];
    }];
}

- (void)updatePin:(NSString *) cardId
           newPin:(NSString *) newPin
     verificationId:(NSString *) verificationId
        oneTimeCode:(NSString *) oneTimeCode
         completion:(STPPinCompletionBlock) completion{
    NSString *endpoint = [NSString stringWithFormat:@"issuing/cards/%@/pin", cardId];
    NSDictionary *parameters = @{
                                 @"verification": @{
                                         @"id": verificationId,
                                         @"one_time_code": oneTimeCode,
                                         },
                                 @"pin": newPin,
                                 };
    [self.keyManager getOrCreateKey:^(STPEphemeralKey * _Nullable ephemeralKey, NSError * _Nullable keyError) {
        if (ephemeralKey == nil) {
            completion(nil, STPPinEphemeralKeyError, keyError);
            return;
        }
        [STPAPIRequest<STPIssuingCardPin *> postWithAPIClient:self.apiClient
                                                     endpoint:endpoint
                                            additionalHeaders:[self.apiClient authorizationHeaderUsingEphemeralKey:ephemeralKey]
                                                  parameters:parameters
                                                deserializer:[STPIssuingCardPin new]
                                                  completion:^(
                                                               STPIssuingCardPin *details,
                                                               __unused     NSHTTPURLResponse *response,
                                                               NSError *error) {
                                                      // Find if there were errors
                                                      if (details.error != nil) {
                                                          NSString* code = details.error[@"code"];
                                                          if ([@"api_key_expired" isEqualToString:code]) {
                                                              completion(nil, STPPinEphemeralKeyError, error);
                                                          } else if ([@"expired" isEqualToString:code]) {
                                                              completion(nil, STPPinErrorVerificationExpired, nil);
                                                          } else if ([@"incorrect_code" isEqualToString:code]) {
                                                              completion(nil, STPPinErrorVerificationCodeIncorrect, nil);
                                                          } else if ([@"too_many_attempts" isEqualToString:code]) {
                                                              completion(nil, STPPinErrorVerificationTooManyAttempts, nil);
                                                          } else if ([@"already_redeemed" isEqualToString:code]) {
                                                              completion(nil, STPPinErrorVerificationAlreadyRedeemed, nil);
                                                          } else {
                                                              completion(nil, STPPinUnknownError, error);
                                                          }
                                                          return;
                                                      }
                                                      completion(details, STPPinSuccess, nil);
                                                  }];
    }];
}

@end

File: Stripe/STPPinManagementService.swift
//
//  STPAPIClient+PinManagement.h
//  Stripe
//
//  Created by Arnaud Cavailhez on 4/29/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

import Foundation
import UIKit

/// STPAPIClient extensions to manage PIN on Stripe Issuing cards
public class STPPinManagementService: NSObject {
    /// The API Client to use to make requests.
    /// Defaults to STPAPIClient.shared
    @objc public var apiClient: STPAPIClient = STPAPIClient.shared

    /// Create a STPPinManagementService, you must provide an implementation of STPIssuingCardEphemeralKeyProvider
    @objc
    public init(keyProvider: STPIssuingCardEphemeralKeyProvider) {
        super.init()
        keyManager = STPEphemeralKeyManager(
            keyProvider: keyProvider as Any, apiVersion: STPAPIClient.apiVersion,
            performsEagerFetching: false)
    }

    /// Retrieves a PIN number for a given card,
    /// this call is asynchronous, implement the completion block to receive the updates
    @objc
    public func retrievePin(
        _ cardId: String,
        verificationId: String,
        oneTimeCode: String,
        completion: @escaping STPPinCompletionBlock
    ) {
        let endpoint = "issuing/cards/\(cardId)/pin"
        let parameters = [
            "verification": [
                "id": verificationId,
                "one_time_code": oneTimeCode,
            ]
        ]
        keyManager?.getOrCreateKey({ ephemeralKey, keyError in
            if ephemeralKey == nil {
                completion(nil, .ephemeralKeyError, keyError)
                return
            }

            if let ephemeralKey = ephemeralKey {
                APIRequest<STPIssuingCardPin>.getWith(
                    self.apiClient,
                    endpoint: endpoint,
                    additionalHeaders: self.apiClient.authorizationHeader(using: ephemeralKey),
                    parameters: parameters
                ) { details, _, error in
                    // Find if there were errors
                    if details?.error != nil {
                        let code = details?.error?["code"] as? String
                        if "api_key_expired" == code {
                            completion(nil, .ephemeralKeyError, error)
                        } else if "expired" == code {
                            completion(nil, .errorVerificationExpired, nil)
                        } else if "incorrect_code" == code {
                            completion(nil, .errorVerificationCodeIncorrect, nil)
                        } else if "too_many_attempts" == code {
                            completion(nil, .errorVerificationTooManyAttempts, nil)
                        } else if "already_redeemed" == code {
                            completion(nil, .errorVerificationAlreadyRedeemed, nil)
                        } else {
                            completion(nil, .unknownError, error)
                        }
                        return
                    }
                    completion(details, .success, nil)
                }
            }
        })
    }

    /// Updates a PIN number for a given card,
    /// this call is asynchronous, implement the completion block to receive the updates
    @objc
    public func updatePin(
        _ cardId: String,
        newPin: String,
        verificationId: String,
        oneTimeCode: String,
        completion: @escaping STPPinCompletionBlock
    ) {
        let endpoint = "issuing/cards/\(cardId)/pin"
        let parameters =
            [
                "verification": [
                    "id": verificationId,
                    "one_time_code": oneTimeCode,
                ],
                "pin": newPin,
            ] as [String: Any]
        keyManager?.getOrCreateKey({ ephemeralKey, keyError in
            if ephemeralKey == nil {
                completion(nil, .ephemeralKeyError, keyError)
                return
            }
            if let ephemeralKey = ephemeralKey {
                APIRequest<STPIssuingCardPin>.post(
                    with: self.apiClient,
                    endpoint: endpoint,
                    additionalHeaders: self.apiClient.authorizationHeader(using: ephemeralKey),
                    parameters: parameters
                ) { details, _, error in
                    // Find if there were errors
                    if details?.error != nil {
                        let code = details?.error?["code"] as? String
                        if "api_key_expired" == code {
                            completion(nil, .ephemeralKeyError, error)
                        } else if "expired" == code {
                            completion(nil, .errorVerificationExpired, nil)
                        } else if "incorrect_code" == code {
                            completion(nil, .errorVerificationCodeIncorrect, nil)
                        } else if "too_many_attempts" == code {
                            completion(nil, .errorVerificationTooManyAttempts, nil)
                        } else if "already_redeemed" == code {
                            completion(nil, .errorVerificationAlreadyRedeemed, nil)
                        } else {
                            completion(nil, .unknownError, error)
                        }
                        return
                    }
                    completion(details, .success, nil)
                }
            }
        })
    }

    private var keyManager: STPEphemeralKeyManager?
}
